### PR TITLE
ch4: add handler for SIGUSR1

### PR DESCRIPTION
**Pull request by yfguo**
SIGUSR1 is a signal sent by PMI when there is a process failure
somewhere in the process tree. The handler of SIGUSR1 is set in
MPID_Init. The handler just record the number of SIGUSR1 received. We
check the number of the received SIGUSR1.  If there is a new SIGUSR1, we
will try to get failed groups from PMI and return an error.

### CODE REVIEW

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

File: src/mpid/ch4/src/ch4_globals.c [hunk 1]
```diff
@@ -107,6 +123,56 @@ int MPID_Abort(MPIR_Comm * comm, int mpi_errno, int exit_code, const char *error
...
+    }
+#ifdef USE_PMI2_API
+    {
+        int vallen = 0;
+        pmi_errno =
+            PMI2_KVS_Get(kvsname, PMI2_ID_NULL, "PMI_dead_processes", failed_procs_string,
+                         PMI2_MAX_VALLEN, &vallen);
+        MPIR_ERR_CHKANDJUMP(pmi_errno, mpi_errno, MPI_ERR_OTHER, "**pmi_kvs_get");
+    }
+#else
```
Comment (raffenet) [inline src/mpid/ch4/src/ch4_globals.c]: Please add a PMIX block, as well. Can be a simple `MPIR_Assert(0);` for now.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

File: src/mpid/ch4/src/ch4_globals.c [hunk 2]
```diff
@@ -107,6 +123,60 @@ int MPID_Abort(MPIR_Comm * comm, int mpi_errno, int exit_code, const char *error
...
+     * with the rank, then we need to create the failed group from
+     * something bigger than comm_world. */
+#ifndef USE_PMIX_API
+    pmi_errno = PMI_KVS_Get_my_name(kvsname, MPIDI_KVSAPPSTRLEN);
+    if (pmi_errno != PMI_SUCCESS) {
+        mpi_errno = MPI_ERR_OTHER;
+        MPIR_ERR_POP(mpi_errno);
+    }
+#endif
+#ifdef USE_PMI2_API
```
Comment (wesbland) [inline src/mpid/ch4/src/ch4_globals.c]: No `#endif`?

File: src/mpid/ch4/include/mpidpre.h [hunk 3]
```diff
@@ -489,6 +489,15 @@ MPL_STATIC_INLINE_PREFIX int MPIDI_Type_free_hook(MPIR_Datatype * type);
 #define MPID_INTERCOMM_NO_DYNPROC(comm) \
     (MPIDI_COMM((comm),map).avtid == 0 && MPIDI_COMM((comm),local_map).avtid == 0)
 
+#ifdef HAVE_SIGNAL
+extern void (*prev_sighandler) (int);
+void sigusr1_handler(int sig);
+#endif
+
+extern volatile int sigusr1_count;
+extern int my_sigusr1_count;
```
[the remark this answers is not in the document]
Comment (raffenet) [reply src/mpid/ch4/include/mpidpre.h]: I don't think there's a reason to expose these beyond the device boundary at this point. Maybe we can just put it in the `MPIDI_CH4_global` struct?

[the remark this answers is not in the document]
Comment (raffenet) [reply src/mpid/ch4/include/mpidpre.h]: Which should be renamed to `MPIDI_global` of course, but that name is currently held by the OFI netmod...future hackathon topic.

File: src/mpid/ch4/src/ch4_globals.c (hunk 2 above)
Comment (yfguo) [reply src/mpid/ch4/src/ch4_globals.c]: It supposes to be `#elif` up there, but I put `#if` there.

**Comment by yfguo** [comment]
test:jenkins/ch4/ofi

File: src/mpid/ch4/src/ch4_globals.c [hunk 4]
```diff
@@ -107,6 +116,60 @@ int MPID_Abort(MPIR_Comm * comm, int mpi_errno, int exit_code, const char *error
...
+    {
+        int vallen = 0;
+        pmi_errno =
+            PMI2_KVS_Get(kvsname, PMI2_ID_NULL, "PMI_dead_processes", failed_procs_string,
+                         PMI2_MAX_VALLEN, &vallen);
+        MPIR_ERR_CHKANDJUMP(pmi_errno, mpi_errno, MPI_ERR_OTHER, "**pmi_kvs_get");
+    }
+#elf defined(USE_PMIX_API)
```
Comment (wesbland) [inline src/mpid/ch4/src/ch4_globals.c]: Should be `#elif`?

File: src/mpid/ch4/src/ch4_globals.c (hunk 2 above)
Comment (wesbland) [reply src/mpid/ch4/src/ch4_globals.c]: Should this be an `#elif` as well? Do we expect to use PMIX and PMI2 at the same time?

File: src/mpid/ch4/src/ch4_globals.c (hunk 4 above)
Comment (yfguo) [reply src/mpid/ch4/src/ch4_globals.c]: Sorry, I got a lot typo recently.

**Comment by yfguo** [comment]
The test looks clean. I will fix the typo and formatting issue. Then, it is ready to go.

File: src/mpid/ch4/src/ch4_globals.c (hunk 2 above)
Comment (yfguo) [reply src/mpid/ch4/src/ch4_globals.c]: This one needs to be `#ifdef`. The above section is disabled only when PMIx is used. Both PMI and PMI2 needs that `PMI_KVS_Get_my_name`.

**Comment by yfguo** [comment]
test:jenkins/ch4/all

**Comment by yfguo** [comment]
Hmm, this is not working properly yet.

**Comment by yfguo** [comment]
test:jenkins/ch4/all

**Comment by yfguo** [comment]
test:jenkins/ch4/all

**Comment by yfguo** [comment]
test:jenkins/ch4/ucx
test:jenkins/ch4/ofi

**Comment by yfguo** [comment]
I have reverted to use `volatile int` for the `sigusr1_count`. The reason is that it is only updated in the signal handler. The `progress_test` just read the value. The only problem without atomics is WAR, which is not an issue in this particular case. The `progress_test` will eventually find out that the `sigusr1_count` is changed.

**Comment by yfguo** [comment]
@raffenet The tests are good. The PR is ready to go.

**Pull request merged**